website/static/js/registrationUtils.js

```diff
@@ -0,0 +1,764 @@
+        }
+        return self.draft().updated;
+    });
+    self.lastSaved = ko.computed(function() {
+        var t = self.lastSaveTime();
+        if (t) {
+            return t.toGMTString();
+        } else {
+            return 'never';
+        }
+    });
+
+    self.canRegister = ko.computed(function() {
+        var draft = self.draft();
+        return draft && draft.isApproved;
+    });
+
+    self.iterObject = $osf.iterObject;
+
+    // TODO: better extensions system?
+    self.extensions = {
+        'osf-upload': editorExtensions.Uploader
+    };
+};
+/**
+ * Load draft data into the editor
+ *
+ * @param {Draft} draft
+ **/
+RegistrationEditor.prototype.init = function(draft) {
+    var self = this;
+
+    self.draft(draft);
+    var metaSchema = draft.metaSchema;
+
+    var schemaData = {};
+    if (draft) {
+        schemaData = draft.schemaData || {};
+    }
+
+    var questions = self.flatQuestions();
+    $.each(questions, function(i, question) {
+        var val = schemaData[question.id];
+        if (val) {
+            if (question.type === 'object') {
+                $.each(question.properties, function(prop, subQuestion) {
+                    val = schemaData[question.id][prop];
+                    if (val) {
+                        subQuestion.value(val.value);
+                    }
+                });
+            } else {
+                question.value(val.value);
+            }
+        }
+    });
+    self.currentQuestion(questions.shift());
+};
+/**
+ * @returns {Question[]} flat list of the current schema's questions
+ **/
+RegistrationEditor.prototype.flatQuestions = function() {
+    var self = this;
+
+    return self.draft().metaSchema.flatQuestions();
+};
+/**
+ * Creates a template context for editor type subtemplates. Looks for the data type
+ * in the extension map, and if found instantiates that type's ViewModel. Otherwise
+ * return the bare data Object
+ *
+ * @param {Object} data: data in current editor template scope
+ * @returns {Object|ViewModel}
+ **/
+RegistrationEditor.prototype.context = function(data) {
+    $.extend(data, {
+        save: this.save.bind(this),
+        readonly: this.readonly
+    });
+
+    if (this.extensions[data.type]) {
+        return new this.extensions[data.type](data);
+    }
+    return data;
+};
+/**
+ * Check that the Draft is valid before registering
+ */
+RegistrationEditor.prototype.check = function() {
+    var self = this;
+
+    var proceed = true;
+    $.each(self.flatQuestions(), function(i, question) {
+        if (question.required) {
+            var valid = question.value.isValid();
+            proceed = proceed && valid;
+        }
+    });
+    if (!proceed) {
+
+        bootbox.dialog({
+            title: 'Registration Not Complete',
+            message: 'There are errors in your registration. Please double check it and submit again.',
+            buttons: {
+                success: {
+                    label: 'Ok',
```

Comment:
See style guide for modal language and revise: https://centerforopenscience.github.io/osf-style/ui.html#modals
> When adding modals also be cognizant of these implementation practices:
>    Phrase the words in terms of the action, i.e. don't use "OK" if possible, use "I agree", or other verb.